### DISCUSSION TITLE
SpineOpt: Multi-period investments

### DISCUSSION THREAD

**Discussion by mihlema** [spine-tools/SpineOpt.jl#476] [Ideas]
# SpineOpt: Multi-period investments

This discussion will lay out the current idea for the economic representation of multi-period investments in SpineOpt. For the sake of brevity, we will illustrate the new functionality using the `unit` class. However, the functionality should also be added for `node`s (storages) and `connection`s. Additional information can also be found here: 
[economic representation.pdf](https://github.com/Spine-project/SpineOpt.jl/files/9130501/economic.representation.1.pdf)
[parameter_illustration.xlsx](https://github.com/Spine-project/SpineOpt.jl/files/9130469/parameter_illustration.2.xlsx)
[Spine_invest.pdf](https://github.com/Spine-project/SpineOpt.jl/files/9130471/Spine_invest.3.pdf)

## General thoughts

The changes proposed here are rather fundamental as they impact the investment structure of SpineOpt. I believe @DillonJ proposed to e.g. have this structure in parallel to the existing structure. I am a bit hesitant about this, given the potential overhead of maintenance and keeping consistency - but I could be convinced if we have a good testing strategy. In that case of having "parallel" implementations as a way forward, we should also revisit other fundamental constraints (e.g. `units_on`, `units_available` etc.)

### Capacity transfer factor
The capacity transfer factor describes when a certain technology becomes available and when it will be retired. The capacity transfer factor is a matrix that will map an investment year to the availability of technology depending on the lead-time of the project, the technical lifetime of the project, and the duration of one investment period. 

An illustrative example of this matrix is given below:
|          | **2020** | **2025** | **2030** | **...** |
|----------|:--------:|:--------:|:--------:|:-------:|
| **2020** | 0.25     | 0        | 0        | ...     |
| **2025** | 1        | 0.25     | 0        | ...     |
| **2030** | 1        | 1        | 0.25     | ...     |
| **2035** | 0.75     | 1        | 1        | ...     |
| **2040** | 0        | 0.75     | 1        | ...     |

The capacity_transfer_factor gets generated at the beginning of run_spineopt (after the temporal and stochastic structures are generated).  The capacity transfer factor will be calculated according to the following rules:

$$
\begin{align}
t_{end\\_construction}  =t^{start}\_v +  p^{lead\\_time} (unit=u)
\end{align}
$$

$$
\begin{equation}
  p^{capacity\_transfer\_factor(u,s,t\_v,t)} =
    \begin{cases}
      0, & \text{if } t.end < t^{start}\_v + p^{lead\_time} \text{ (unit=u)} \\
      \max\left(\min\left(1 - \frac{t_{end\_construction} - t^{start}}{\Delta t},1\right),0\right), & \text{if } t^{start}\_v < t_{end\_construction} \leq t^{end} \\
      \max\left(\min\left(\frac{t_{end\_construction} + p^{technical\_lifetime}(unit=u) - t^{start}\_v}{\Delta t},1\right),0\right), & \text{otherwise}
    \end{cases}
\end{equation}
$$
      

The nested map parameter can be accessed as follows:
 `capacity_transfer_factor(unit=u, stochastic_scenario=s, t_vintage=t_v, t=t)` 
Note that the stochastic scenario used for this parameter is based on the stochastic scenario active during the investment period, i.e., `t_v`.

To decide:
- What happens if the lead time is not defined? The current default is to assume that the technology becomes immediately available at the beginning of the investment period
- What happens if technical lifetime is not defined? The current default is to assume that the technology will "live forever" until the end o the modelling horizon
- How should we ensure rolling horizon functionality? Currently, the parameter gets created *before* the rolling loop. Given the current implementation, we would need to "re-write" the parameter on every role, and make sure that the parameter call is updating (using the [] operator in constraints).

### Discounted annuities
The `unit_conversion_to_discounted_annuities` factor translates the overnight costs of investment into discounted (to the `discount_year`) annual payments, distributed over the total lifetime of the investment. Investment payments are assumed to increase linearly over the lead-time, and decrease linearly towards the end of the economic lifetime. This is also illustrated here: 
![image](https://user-images.githubusercontent.com/40030652/178750722-16d8c749-7fa7-47cc-b832-3ead19125d73.png)

For this purpose, we first calculate first the fraction of payment per year (e.g. something like 0.25, 0.5, 0.75,1 over the lead time; 1 for the economic lifetime minus the lead time, and 0.75, 0.5, 0.25 and 0 for the remaining economic lifetime).  Each payment fraction is then multiplied by the discounting factor of the payment year with respect to the discounting year (e.g. start of optimization). 

The sum of these products is then multiplied by the capital recovery factor, which corresponds to the `unit_conversion_to_discounted_annuities`  factor. (The capital recovery factor is the ratio between the annuity and the present value of an investment.) This way, the overnight investment costs in the objective function only need to be multiplied with this factor to be translated into discounted, annual (distributed) payments. The parameter is generated for each possible vintage year (as given by the investment temporal blocks).
### Discounted annuities for decommissioning
Similarly to the distribution of investment costs, we also allow for the discounting of decommissioning costs and their conversion to annuities. In this case, we assume that the `unit_decommissioning_cost`s are equally distributed over the `unit_decommissioning_time`. Hence, the payment fraction (i,e, here 1) of each decommissioning year is multiplied by its corresponding discount factor, and the sum of these products will be multiplied by the capital recovery factor, and saved, for each possible vintage year, in the parameter `unit_decommissioning_conversion_to_discounted_annuities`
### Salvage fraction
As we consider (discounted and annuitized) overnight costs in the objective function, it can happen that the lifetime of a unit exceeds the model horizon. In such cases, the salvage fraction needs to be deducted from the objective function. (Note: Alternatively this could automatically be considered in the calculation of the discounted annuities)

In principle, this means, that the annuities "already paid", which exceed the modelling horizon, are recuperated.
### Discounted durations
The discounted duration is used to discount operational costs within a certain investment period to the discount year (e.g. beginning of the optimization). If milestone years are used for investments, the discounted duration is calculated for each investment period as defined by investment temporal blocks, otherwise, it will be calculated on a yearly basis.

$$
\begin{align}
`intra\_period\_discounting\_factor(t)` & = \sum_{\delta t' \in 0Y:1Y:\Delta t} \frac{1}{(1+i)^{\delta t'}}\\
`dicounted\_duration(t)` & = \frac{1}{(1+i)^{t-t_{discount}}}\cdot `intra\_period\_discounting\_factor(t)`\\
\end{align}
$$

As an example, if we use milestone years, and have an investment period of 5 years. With a discount rate of 0.05, the discounted duration for 2030 (assuming a discount year 2020) corresponds to 2.79. Note, that when using milestone years, the discounted duration will reflect the duration of the period, and hence the weight of the operational temporal blocks (e.g. representative day) should correspond to the weight within one year (e.g. 1 temporal_block that is repeated 7 times throughout the milestone year will have a weight of 7)

### Technology specific discount factors
The technology specific discount factor can be used, if e.g. investments in a certain technology are particularly risky. In this case, the annuitiesed, discounted investment costs are multiplied with the capital recovery factor associated with the technology-specific discount rate, and divided by the capital recovery factor associated with the common discount rate. (The default value for this parameter is 1.)
## Relevant variables
- `units_available`: Operational variables; unit's available for dispatch
- `units_invested`: Number of investment decision taken at a specific point in time
- `units_invested_available`: Number of available invested units (i.e., sum of all invested units minus 
decommissioned units minus mothballed units)
- `units_invested_available_vintage`: Number of available invested units, with respect to a certain vintage year; i.e. a specific year of `units_invested` 
- `units_invested_state`: Sum of invested units, that have not been decommissioned (i.e., the sum of all invested units minus decommissioned)
- `units_invested_state_vintage`: Sum of invested units, that have not been decommissioned, with respect to a certain vintage year; i.e. a specific year of `units_invested` 
- `units_mothballed_state`: Sum of all units that have been mothballed, i.e. units are *temporarily* not operated, but can be brought back online at a later stage
- `units_mothballed_state_vintage`: Sum of all units that have been mothballed with respect to a certain vintage year
- `units_decommissioned`: Sum of units that have been decommissioned
- `units_decommissioned_vintage`: Sum of units that have been decommissioned with respect to a certain vintage year
- `units_early_decommissioned_vintage`: Sum of units that have been *early* decommissioned with respect to a certain vintage year; i.e. before the end of their technical lifetime
## New constraints:
Upper bound on `units_invested_available`:

$$
\begin{align}
  & `units\_invested\_state\_vintage` (u, t\_v, t) \\
    & = \\
    & `capacity\_transfer\_factor`(u, t_v,t) \cdot \\
    & (`units\_invested`(u, t_v) - \sum_{t'= t-t_{`technical\_lifetime`}}^{t} `units\_early\_decomissioned`(u, t_v,t') )
\end{align}
$$

$$
\newline
$$

$$
\begin{align}
    `units\_invested\_state\_vintage`(u, t,t_v) >=0
\end{align}
$$

$$
\newline
$$

$$
\begin{align}
    & `units\_invested\_state`(u, t) \\ 
    & == \\
    & \sum_{t_v = t - t_{`technical\_lifetime`}}^{t} `units\_invested\_state\_vintage`(u, t,t_v)
\end{align}
$$

Mothballing stuff:

$$
\begin{align}
    & `units\_mothball\_state\_vintage`(u, t,t_v)\\
    & ==\\
    & \sum_{t - t_{`technical\_lifetime`}}^{t}+ `units\_mothballed\_vintage`(u, t,t_v)\\ 
    & - \sum_{t - t_{`technical\_lifetime`}}^{t} `units\_demothballed\_vintage`(u, t,t_v)
\end{align}
$$

$$
\newline
$$
$$
\begin{align}
     `units\_mothball\_state\_vintage`(u, t,t_v) >= 0
\end{align}
$$

Add constraints that links available, mothballed and invested state units:

$$
\begin{align}
    & `units\_invested\_available\_vintage`(u, t, t_v) \\
    & == \\
    &  `units\_invested\_state\_vintage`(u, t, t_v)  \\
    & - `units\_mothball\_state\_vintage`(u, t,t_v)
\end{align}
$$

$$
\newline
$$

$$
\begin{align}
    & `units\_invested\_available`(u, t) \\ 
    & == \\
    & \sum_{t_v = t - t_{`technical\_lifetime`}}^{t} `units\_invested\_available\_vintage`(u, t,t_v)
\end{align}
$$

$$
\newline
$$

$$
\begin{align}
    & `units\_decommissioned\_vintage`(u, t_v, t) >=\\
    & - `units\_invested\_state\_vintage`(u, t, t_v) \\
    & + `units\_invested\_state\_vintage`(u, t_{before}, t_v) 
\end{align}
$$

## Adaptions objective terms:
- all operational cost terms are multiplied with the discounted duration factor\\
- investment costs are multiplied with discounted annuties conversion factor and the technological discount factor and (1-salvage fraction)
- decommissioned costs are multiplied with discounted decommissioning annuities (Note: we could also add a salvage fraction for decommissioning in the future)
- mothballing costs are assumed to happen instantaniously, and thus only multiplied with the discount fctor of the mothballing year

## Implementation status
Implementation of this feature is ongoing. Feel free to check out [this branch](https://github.com/Spine-project/SpineOpt.jl/tree/economic_representation) I'm working on.
### Look back time
One thing I noticed is that it can take really long if we create the operational history for e.g. 40 Years. Therefore we now have a short- and a long history depending on the variable type. By default, variables have a long history
### Vintage years
We now have vintage years, so in the future, we could also include vintage of operation etc. However, for that to really work I think we'll have to think a bit more careful about stochastic vintage etc.
### Apriori parameter calculation
While it is convenient to have the parameter generation "outside" the core-building of the model, i.e. these parameters are calculated before the constraints etc are build, it also comes with some disadvantages. In some cases when the temporal structure is constant, it might be more convenient to have a small function, rather than storing everything in (nested) maps. But for the moment this in turn would be kind of tricky when we move to time-varying structures...
### A word on stochastic structures
I *tried* to allow for stochastic structures as much as possible. However, things very quickly become really messy, especially as temporal and stochastic structures are linked, but this link is not always immediately accessible. For example, when there are stochastic parameters related to the investment structure, that also have an impact on operations (e.g. discount rate and operational costs), things become quite funky.
In the long run, I think as the next big improvement to SpineOpt, we should think about a way to refurbish our tempo-stoch structure a bit. 

@Spine-project/spineopt-developers

**Comment by Tasqu** [comment]
>In the long run, I think as the next big improvement to SpineOpt, we should think about a way to refurbish our tempo-stoch structure a bit.

I agree. While quite flexible, the implementation of the stochastic and temporal structures is quite complicated. I'd go as far as to say that it's a barrier for new users and further development. Streamlining and simplifying it could be a really big help, in my opinion. It would be nice to have the time and resources to try and tackle this issue at some point.

Personally, I'd also like to see if we could separate handling the stochastic-temporal structures into a dedicated module, which would then be used by SpineOpt. That way, the work *(if done properly)* could potentially serve other models as well.

**Reply by DillonJ** [reply]
Nice write-up @mihlema 

On the point of refurbishing the tempo-stochastic structure - I definitely agree that we could make some user-comfort improvements to the usability of the system. However, I would say that in my opinion the flexibility of the current system is findamental to the power of SpineOpt. We are currently using all the functionality and in ways we never imagined when it was conceived originally - so I think it's important that whatever we do, we don't lose any of this functionality. 

Also, I have found that when doing anything semi-advanced with the temporal and stochastic structure - you really need to understand quite well how it's working under the hood and there are plenty of pitfalls - particularly the interaction with default temporal blocks etc. However, when you've worked with the system a lot, I think it's actually pretty simple - but it takes a while for a new user to get into the SpineOpt frame of mind, and most would probably give up before they got there. I think some very simple, tutorial systems could help enormously with that - I think that's the single biggest missing feature for new users... a guide to build up a small, simple system, step by step, introducing new concepts and features incrementally.

Another point relating to the tempo-stoch structure is the issue of maps - we find that it is impossible to do anything with maps without writing code - this would be a big barrier for many users - we need some tools to manage maps as inputs and outputs.

Some specific things that I think could help:
 - As above, a tutorials to build up simple systems introducing features and functionalities incrementally
 - Expand the model templates to include more exotic structures
 - More data checks on the integrity of the temporal/stochastic structure
 - Perhaps additional tools, akin to what we have to manage scenarios and alternatives, to manage the temporal/stochastic structure of a model

**Reply by Tasqu** [reply]
>However, I would say that in my opinion the flexibility of the current system is findamental to the power of SpineOpt. We are currently using all the functionality and in ways we never imagined when it was conceived originally - so I think it's important that whatever we do, we don't lose any of this functionality.

I completely agree. Just to clarify, by *simplifying* I meant trying to see if the functionality could be wrapped up nicer. All the different index sets and deducing the stochastic path indexing for the constraints is a bit intimidating at the moment, so I wonder if it could be made clearer somehow.  Maybe *clarifying* was the word I was looking for 😅

I didn't spend a whole lot of time revising the stochastic structure code after I got it working during the project *(and I was still somewhat new to Julia)*, so I imagine there could be a lot of room for improvement.

**Comment by mihlema** [comment]
Combing back to this discussion, and taking some notes on possible improvements.
As described above, the new implementation would introduce vintage to all investment variables, which could be used lateron to e.g. also use degradation over time etc. However, an alternative route could be to realize this from the dataside.

The idea is to "force" vintage units from the data side. E.g. consider 2 units `unit_2020` and `unit_2021` (of the same type), for both we define a TimeVarying `candidate_units` parameter. `unit_2020` and `unit_2021` can only be invested in in the year 2020 and 2021, respectively. Degradation over time could then also be added with a time-varying parameter for each of these units (e.g. `2020: 0.8, 2021: 0.76, ...]`. 

Codewise this is probably a bit more in-line with what we already have, and the impact on existing code is less pronounced. Datawise we probably don't want to have e.g. 50 units for 50 investment periods, but I guess we can do something about that with preprocessing.
